### PR TITLE
Added string escaping support throughout the expr language, and fixed…

### PR DESCRIPTION
**Pull request by tpoterba**
… annotation gen to allow for sets to be randomly generated.  Split out from #406

### CODE REVIEW

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
```
BUILD SUCCESSFUL
```

File: src/main/scala/org/broadinstitute/hail/check/Arbitrary.scala [hunk 1]
```diff
@@ -9,6 +9,8 @@ object Arbitrary {
   implicit def arbBoolean: Arbitrary[Boolean] = new Arbitrary(
     Gen.oneOf(true, false))
 
+  implicit def arbByte: Arbitrary[Byte] = new Arbitrary(Gen.arbByte)
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/check/Arbitrary.scala]: I would say `Gen.oneOf(Byte.MinValue, -1, 0, 1, Byte.MaxValue)` or an arbitrary byte.

File: src/main/scala/org/broadinstitute/hail/check/Gen.scala [hunk 2]
```diff
@@ -214,11 +214,17 @@ object Gen {
...
+    p.rng.getRandomGenerator.nextInt().toByte
+  }
 
-  def arbInt: Gen[Int] = Gen { p => p.rng.getRandomGenerator.nextInt() }
+  def arbInt: Gen[Int] = Gen { p =>
+    p.rng.getRandomGenerator.nextInt()
+  }
 
-  def arbLong: Gen[Long] = Gen { p => p.rng.getRandomGenerator.nextLong() }
+  def arbLong: Gen[Long] = Gen { p =>
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/check/Gen.scala]: Get rid of all of these.  We should use `arbitrary[T]`.

File: src/main/scala/org/broadinstitute/hail/expr/Parser.scala [hunk 3]
```diff
@@ -304,21 +295,47 @@ object Parser extends JavaTokenParsers {
...
+            escapeUnprintable(toSearch.charAt(m.start).toString)
+          }'""")
+      }
+
+  override def stringLiteral: Parser[String] =
+    ("\"" + """([^"'\p{Cntrl}\\]|\\[\\"'bfnrt]|\\u[a-fA-F0-9]{4})*""" + "\"").r ^^
+      (s => unescapeString(s.substring(1, s.length - 1))) |
+      withPos(("\"" + ".*".r + "\"").r) ^^ { r =>
+        val toSearch = r.x.substring(1, r.x.length - 1)
+        val matches = """[\p{Cntrl}]|[^\\][\\'"bfnrt`]""".r.findFirstMatchIn(toSearch)
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/Parser.scala]: Shouldn't this be ```\\[^\\'"bfnrt`]```?

Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/Parser.scala]: I think I'm inclined to make this more flexible.  Accept strings with control characters.  I don't see any reason not to, except that C (and its syntactic descents) don't, although that changed in C++11 with raw (verbatim) string literals.

http://stackoverflow.com/questions/3093632/why-must-c-c-string-literal-declarations-be-single-line


You can just match `"([^"]|\\.)*"` and catch an exception from `unescapeString` for invalid escape characters.  But we should still escape unprintable characters when escaping on output.

Also, then you can match just `"` and generate an error for unterminated string literal.

Same for backtick identifier.

File: src/main/scala/org/broadinstitute/hail/expr/Type.scala [hunk 4]
```diff
@@ -97,6 +108,10 @@ abstract class Type extends BaseType {
     sb.append(toString)
   }
 
+  def compact(sb: StringBuilder, printAttrs: Boolean = false) {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/Type.scala]: I would have made `compact: Boolean` an flag to `pretty`.  This way is a bit redundant.

File: src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    var sz: Int = 0
+    sz = str.length
+    var i: Int = 0
+    while (i < sz) {
+      {
+        val ch: Char = str.charAt(i)
+        if (ch > 0xfff) {
+          sb.append("\\u" + hex(ch))
+        }
+        else if (ch > 0xff) {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala]: There are a ton of dangling elses in this file.

File: src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+        }
+      }
+      i += 1
+    }
+    sb.result()
+  }
+
+  def escapeBackticked(str: String): String = escapeBackticked(str, new StringBuilder(capacity = str.length * 2))
+
+  def escapeBackticked(str: String, sb: StringBuilder): String = {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala]: As far as I can tell, this duplicates `escapeString` except with one extra case for ```\````.  I'd just add a `escapeBacktick: Boolean` flag argument.

File: src/test/scala/org/broadinstitute/hail/methods/ExprSuite.scala [hunk 7]
```diff
@@ -174,6 +176,15 @@ class ExprSuite extends SparkSuite {
     assert(eval[Int]("""iset.max""").contains(2))
     assert(eval[Int]("""iset.sum""").contains(3))
 
+    assert(eval[String](""" "\t\t\t" """).contains("\t\t\t"))
+    assert(eval[String](""" "\"\"\"" """).contains("\"\"\""))
+    assert(eval[String](""" "```" """).contains("```"))
+    val badInput1 = intercept[FatalException](eval[String](" \"aaa\t\t\taaa\" "))
+    assert(badInput1.getMessage.contains("invalid character in string literal: `\\t'"))
```
Comment (cseed) [inline src/test/scala/org/broadinstitute/hail/methods/ExprSuite.scala]: Use `interceptFatal` in `TestUtils`.

**Comment by cseed** [comment]
Back to you.  Looks great, all small stuff.

File: src/main/scala/org/broadinstitute/hail/expr/Parser.scala (hunk 3 above)
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/Parser.scala]: So tabs and \t on input should be functionally equivalent?

**Comment by tpoterba** [comment]
Back to you.

**Comment by tpoterba** [comment]
jenkins

**Comment by jigold** [comment]
jenkins

File: src/main/scala/org/broadinstitute/hail/check/Prop.scala [hunk 8]
```diff
@@ -12,7 +12,7 @@ abstract class Prop {
       if (!System.getProperty("hail.randomize", "false").toBoolean)
         System.getProperty("hail.seed", "1").toInt
       else
-        Gen.arbInt.sample()
+        Gen { p => p.rng.getRandomGenerator.nextInt() }.sample()
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/check/Prop.scala]: What?  This is crazy code.  Just do `Random.nextInt()`.

File: src/main/scala/org/broadinstitute/hail/check/Prop.scala [hunk 9]
```diff
@@ -27,7 +27,7 @@ abstract class Prop {
       if (!random && !System.getProperty("hail.randomize", "false").toBoolean)
         seed.getOrElse(System.getProperty("hail.seed", "1").toInt).toLong
       else
-        Gen.arbInt.sample()
+        Gen { p => p.rng.getRandomGenerator.nextInt() }.sample()
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/check/Prop.scala]: Same here.

File: src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+              sb += '\\'
+              sb += 'f'
+            case '\r' =>
+              sb += '\\'
+              sb += 'r'
+            case _ =>
+              if (ch > 0xf) {
+                sb.append("\\u00" + hex(ch))
+              }
+              else {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala]: Stil dangling elses.

File: src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+                sb.append("\\u000" + hex(ch))
+              }
+          }
+        }
+        else {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala]: Also here.

File: src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+    escapeString(str, new StringBuilder(capacity = str.length * 2), backticked)
+
+  def escapeString(str: String, sb: StringBuilder, backticked: Boolean): String = {
+    sb.clear()
+
+    var sz: Int = 0
+    sz = str.length
+    var i: Int = 0
+    while (i < sz) {
+      {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala]: Extra open brace?

File: src/main/scala/org/broadinstitute/hail/expr/Parser.scala [hunk 13]
```diff
@@ -304,21 +295,47 @@ object Parser extends JavaTokenParsers {
...
+  def structField: Parser[(String, AST)] = (stringLiteral ~ ":" ~ expr) ^^ { case id ~ _ ~ ast => (id, ast) }
 
   def indexStruct: Parser[(String, AST)] = "index" ~ "(" ~ expr ~ "," ~ identifier ~ ")" ^^ {
     case (_ ~ _ ~ ast ~ _ ~ key ~ _) => (key, ast)
     //    case (_ ~ _ ~ ast ~ _ ~ key ~ _) => (key.substring(1, key.length - 1), ast)
   }
 
+  def backtickLiteral: Parser[String] =
+    ("`" + """([^\\`]|\\[\\bfnrt`]|\\u[a-fA-F0-9]{4})*""" + "`").r ^^
+      (s => unescapeString(s.substring(1, s.length - 1))) |
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/Parser.scala]: You didn't like my proposal?

File: src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+
+import org.broadinstitute.hail.Utils._
+
+object StringEscapeUtils {
+
+  def hex(ch: Char): String = Integer.toHexString(ch).toUpperCase(Locale.ENGLISH)
+
+  def escapeUnprintable(str: String): String = escapeUnprintable(str, new StringBuilder(capacity = str.length * 2))
+
+  def escapeUnprintable(str: String, sb: StringBuilder): String = {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala]: I'd either make this take `Char` (the only way it is called) and then call it from `escapeString`, or delete it and just use `escapeString` when reporting unprintable characters.

File: src/main/scala/org/broadinstitute/hail/expr/Parser.scala (hunk 13 above)
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/Parser.scala]: we lose the position if that happens.  I think it's worth the somewhat cluttered code to get a carat where the problem is, but if you don't agree I'll remove it.

File: src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+              sb += 't'
+            case '\f' =>
+              sb += '\\'
+              sb += 'f'
+            case '\r' =>
+              sb += '\\'
+              sb += 'r'
+            case _ =>
+              if (ch > 0xf) {
+                sb.append("\\u00" + hex(ch))
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala]: On reflection, I think non-printable characters < 0x7f should be printed in octal: `\xxx`, where xxx is the 3-digit octal representation.

File: src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+              }
+              else {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/utils/StringEscapeUtils.scala]: Another else.

File: src/test/scala/org/broadinstitute/hail/methods/FilterSuite.scala [hunk 17]
```diff
@@ -168,8 +168,11 @@ class FilterSuite extends SparkSuite {
       .mapAnnotations((v, va, gs) => i(va, Some(1000)))
       .copy(vaSignature = sigs)
     val state = SplitMulti.run(State(sc, sqlContext, vds), Array.empty[String])
-    val s2 = FilterVariantsExpr.run(state, Array("--keep", "-c", "va.`weird name \t test` > 500"))
+    val s2 = FilterVariantsExpr.run(state, Array("--keep", "-c", "va.`weird name \\t test` > 500"))
     assert(s2.vds.nVariants == vds.nVariants)
+
+    val e = intercept[FatalException](FilterVariantsExpr.run(state, Array("--keep", "-c", "va.`bad\\input` == 5")))
```
Comment (cseed) [inline src/test/scala/org/broadinstitute/hail/methods/FilterSuite.scala]: Use `interceptFatal` in `TestUtils`.

File: src/main/scala/org/broadinstitute/hail/expr/Parser.scala (hunk 3 above)
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/Parser.scala]: Yep!  Why not?

File: src/main/scala/org/broadinstitute/hail/expr/Parser.scala (hunk 13 above)
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/Parser.scala]: You don't lose it, you just have to compute it from the string by counting newlines and the position from the last newline.

Or do you mean lose it downstream, if we want to generate messages internal to the string?  That's true.  Once we unescape, we don't know which newlines were real and which were `\n`.  Although I don't think we do that anywhere.

Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/Parser.scala]: I think I must not have understood your proposal.  I thought you wanted the error-catching to be done in `unescapeString`, but then we lose our positioned `parseError`.  Unless we're doing checking in both places, as we are here.

Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/Parser.scala]: Ah!  That part of the proposal.  Yes, you're right, you need to check in both places.

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
Back to you.  I didn't do the octal changes or the desired error handling, but I made an issue to do that when I have a little more time.

**Pull request merged**